### PR TITLE
[FLINK-17452][hive] Support creating Hive tables with constraints

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

Support creating hive tables with constraints.

## Brief change log

  - Add shim method `createTableWithConstraints`.
  - When creating a hive table, i.e. non-generic, get PK and NOT NULL from table schema and call `createTableWithConstraints`.
  - Add test case.

## Verifying this change

Manually run the new test with hive-3.1.1 profile.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Comment by lirui-apache** [comment]
@JingsongLi Please take a look. Thanks.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 1]
```diff
@@ -374,8 +375,30 @@ public void createTable(ObjectPath tablePath, CatalogBaseTable table, boolean ig
...
+
 		try {
-			client.createTable(hiveTable);
+			if (pkConstraint != null || !notNullCols.isEmpty()) {
+				// for now we just create constraints that are DISABLE, NOVALIDATE, RELY
+				Byte[] pkTraits = new Byte[pkConstraint == null ? 0 : pkConstraint.getColumns().size()];
+				Arrays.fill(pkTraits, HiveTableUtil.relyConstraint((byte) 0));
+				Byte[] nnTraits = new Byte[notNullCols.size()];
+				Arrays.fill(nnTraits, HiveTableUtil.relyConstraint((byte) 0));
+				client.createTableWithConstraints(hiveTable, hiveConf,
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:397]: Can client do more things to include upper couple of lines?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV310.java [hunk 2]
```diff
@@ -239,4 +241,54 @@ public Writable hivePrimitiveToWritable(Object value) {
 		}
 		throw new FlinkHiveException("Unsupported primitive java value of class " + value.getClass().getName());
 	}
+
+	@Override
+	public void createTableWithConstraints(IMetaStoreClient client, Table table, Configuration conf,
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV310.java:246]: Too many parameters should be one parameter one line, both call and declaration are like this.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java (hunk 1 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:397]: The client is meant to be just a wrapper of `IMetaStoreClient`. So I think it's more natural to keep such logic in `HiveCatalog`. We can add util method if that logic has to be repeated somewhere else.

**Comment by JingsongLi** [comment]
Thanks for updating. Merged.

**Pull request merged**